node: e2e: device plugins: Deflake e2e tests

swatisehgal opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind cleanup


#### What this PR does / why we need it:
Some node e2e tests check for expected number of pods running
on the node to verify the correct state of that node after running
 test scenarios. An example of such a check is in the device plugin
 end to end test here:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/device_plugin_test.go#L189-L190.
    
 If the node is not left in a clean state after an e2e test finishes
 running, it can lead to flaky tests because the node might have
 unexpected pods running on the node.
    
 In order to avoid that, we make sure that the test pods are
 cleaned up after the test runs in case log rotation e2e tests.
  
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #112612

#### Special notes for your reviewer:
Related PR that was used to debug the issue: https://github.com/kubernetes/kubernetes/pull/113165

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```
<!--
#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:


This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>

```docs

```
-->

swatisehgal committed:
node: e2e: ensure log rotation pod is deleted after test

Some node e2e tests check for expected number of pods running
on the node to verify the correct state of that node after running
test scenarios. An example of such a check is in the device plugin
end to end test here: [1].

If the node is not left in a clean state after an e2e test finishes
running, it can lead to flaky tests because the node might have
unexpected pods running on the node.

In order to avoid that, we make sure that the test pods are
cleaned up after the test runs.

[1]: https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/device_plugin_test.go#L189-L190

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: e2e: ensure clean cluster state before e2e tests are run

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

oomichi commented on test/e2e/framework/expect.go:
```diff
@@ -90,3 +90,23 @@ func ExpectHaveKey(actual interface{}, key interface{}, explain ...interface{})
 func ExpectEmpty(actual interface{}, explain ...interface{}) {
 	gomega.ExpectWithOffset(1, actual).To(gomega.BeEmpty(), explain...)
 }
+
```
Why do we need to add the following functions?
Those functions are not used in this pull request.

swatisehgal replied:
In my initial attempt to resolve this flake, I was going to made amendments to the device plugin e2e test by checking that the number of pods are greater than 2 instead of strict equality as in some cases there was a log rotation pod unexpectedly still running on the node. I later pivoted to an approach of making sure that the log rotation pods are cleaned up properly. 

In spite of the change of approach, I thought it would be useful to leave the functions that I added as `expect.go` is a wrapper around gomega matcher and we could benefit from some additional functions that provide us the ability to perform more granular comparisons in the future. 

I deliberately kept it as a separate commit in case it is considered unnecessary and needs to be removed.

xmcqueen replied:
This looks like a nice and sensible cleanup step for the device plugin test.  I see that one of the new functions is used in this test: `ExpectEqual`.  It does seem a bit out-of-place to put the other new functions in this PR, but I do like them.  I am in favor of keeping them and getting this PR going to reduce the flake count.

swatisehgal commented on test/e2e/framework/expect.go:
```diff
@@ -90,3 +90,23 @@ func ExpectHaveKey(actual interface{}, key interface{}, explain ...interface{})
+	gomega.Expect(actual).Should(gomega.BeNumerically(">=", extra))
+}
+
+// ExpectGreaterThan expects actual is great than extra
```
Thanks, will fix.

SergeyKanzhelev commented:
/priority important-longterm
/triage accepted
/assign @xmcqueen

SergeyKanzhelev commented:
/triage accepted

matthyx commented:
waiting for @xmcqueen enrolment to assign him...

matthyx commented:
/assign @xmcqueen

xmcqueen commented:
/lgtm

SergeyKanzhelev commented:
/assign

aojea commented on test/e2e_node/container_log_rotation_test.go:
```diff
@@ -72,10 +73,20 @@ var _ = SIGDescribe("ContainerLogRotation [Slow] [Serial] [Disruptive]", func()
+			logRotationPod = e2epod.NewPodClient(f).CreateSync(pod)
+		})
+
+		ginkgo.AfterEach(func() {
```
we are moving cleanups to use `ginkgo.DeferCleanup`, see https://github.com/kubernetes/kubernetes/pull/114425 as an example

swatisehgal replied:
Thanks, updated.

swatisehgal commented:
/cc @pohly @SergeyKanzhelev 

for approval

aojea commented:
/assign @pohly

pohly commented on test/e2e/framework/expect.go:
```diff
@@ -90,3 +90,23 @@ func ExpectHaveKey(actual interface{}, key interface{}, explain ...interface{})
+// ExpectLessThanOrEqualTo expects actual is less than or equal to extra
+func ExpectLessThanOrEqualTo(actual interface{}, extra interface{}) {
+	gomega.Expect(actual).Should(gomega.BeNumerically("<=", extra))
+}
```
Please use gomega directly instead of extending `expect.go`

swatisehgal replied:
Ack, removed the commit where additional methods were added in order to encourage the use of gomega directly.

swatisehgal committed:
node: e2e: address review comments (2022/12/12)

- use `ginkgo.DeferCleanup` instead of clean up in the AfterEach block
- encourage use of ginkgo by not extending expect.go

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

pohly commented on test/e2e_node/container_log_rotation_test.go:
```diff
@@ -72,10 +73,15 @@ var _ = SIGDescribe("ContainerLogRotation [Slow] [Serial] [Disruptive]", func()
 			ginkgo.By("get container log path")
-			framework.ExpectEqual(len(pod.Status.ContainerStatuses), 1)
-			id := kubecontainer.ParseContainerID(pod.Status.ContainerStatuses[0].ContainerID).ID
+			framework.ExpectEqual(len(logRotationPod.Status.ContainerStatuses), 1)
```
Please add a message explaining what gets compared here.

swatisehgal replied:
Ack, Updated.

pohly commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -103,6 +105,18 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
 			}, time.Minute, time.Second).Should(gomega.BeTrue())
 
+			v1alphaPodResources, err = getV1alpha1NodeDevices()
+			framework.ExpectNoError(err)
```
Please add a message explaining which operation failed here.

swatisehgal replied:
Fixed, Thanks!

pohly commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -103,6 +105,18 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
+			framework.ExpectNoError(err)
+
+			v1PodResources, err = getV1NodeDevices()
+			framework.ExpectNoError(err)
```
Please add a message explaining which operation failed here.

pohly replied:
Would also be a good enhancement for other such calls.

swatisehgal replied:
Updated, Thanks.

swatisehgal replied:
I would be inclined to split other such fixes to a separate PR to avoid expanding the scope of this PR.

pohly commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -103,6 +105,18 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
+			// Before we run the device plugin test, we need to ensure
+			// that the cluster is in a clean state and there are no
+			// pods running on this node.
+			framework.ExpectEqual(len(v1alphaPodResources.PodResources), 0)
```
```suggestion
			gomega.Expect(v1alphaPodResources.PodResources).To(gomega.BeEmpty(), "should have no pod resources")
```

If this goes wrong, it will print the unexpected pod resources.

swatisehgal replied:
Fixed, Thanks.

pohly commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -103,6 +105,18 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
+			// that the cluster is in a clean state and there are no
+			// pods running on this node.
+			framework.ExpectEqual(len(v1alphaPodResources.PodResources), 0)
+			framework.ExpectEqual(len(v1PodResources.PodResources), 0)
```
Same here.

swatisehgal replied:
fixed, Thanks.

swatisehgal committed:
node: e2e: Add descriptive messages for operation/error checks

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

pohly reviewed (approved):
/lgtm

swatisehgal commented:
/retest

aojea commented:
/lgtm

@SergeyKanzhelev for approval

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

thank you!
